### ISSUE TITLE
As air quality expert I would like that the automatic computation of the emissions done by an AI model of the Open Data Hub is also applied to a new type of traffic measurement technology, i.e. cameras

### ISSUE THREAD

**Issue by rcavaliere** [noi-techpark/bdp-elaborations#38]

**Comment by rcavaliere** [comment]
@clezag before start this activity, please do the following:

- upload the new emission factors for 2023 (see attached file), as documented here: https://github.com/noi-techpark/bdp-elaborations/tree/main/pollution_v2 

```
Update "parco circolante"

"Parco circolante" stands for the configuration containing the estimate of the distribution of the types of car moving on the considered road.

The folder pollution_v2/src/pollution_connector/model/input contains a dedicated CSV file for each year (e.g. fc_info_2018.csv) with a default one containing default values in case of processing a year with the corresponding file missing (fc_info.csv).

When processing data for a specific year, the DAG looks for the file fc_<year>.csv: if found, the file is used, otherwise the system will use the default file fc.csv (e.g. with files fc_2022.csv, fc_2023.csv and fc.csv, when processing data for the year 2023 the system will use fc_2023.csv and when processing data for 2020 and/or 2024 the system will use fc.csv. No configuration needs to be changed, just add the updated file, clean already processed records and let the DAGs run.

To update the "parco circolante" for a specific year, just add the corresponding file (and update the default file, if appropriate) and then [reset the Airflow cache](https://github.com/noi-techpark/bdp-elaborations/tree/main/pollution_v2#Reset-Airflow-Redis-cache) for the previously computed data.
```

Clean previously updated data on ODH and run again the corresponding DAGs

[fc_info_2023.csv](https://github.com/user-attachments/files/17943548/fc_info_2023.csv)

- delete all the emissions already calculated since 1.1.2023, since the elaborations will then consider the emission factors of 2023

The emission calculation for the cameras should start from 10.07.2023 (the day when the data related to the new cameras where released by A22)

**Comment by clezag** [comment]
@rcavaliere The import is currently running in testing with all the changes in place. 
Over the weekend the docker server unfortunately crashed due to the heavy load, so we don't have all the backfilled data yet.

But it should already be possible to check the sensor type and parameter changes. 
In the database I see some camera stations that now have emission measurements.
```sql
select s.name, d.json->>'sensor_type', t.cname, m.timestamp from station s
join measurement m on m.station_id  = s.id
join metadata d on d.id = s.meta_data_id 
join type t on t.id = m.type_id 
where t.cname like '%-emissions'
and m.timestamp > to_date('01012023', 'DDMMYYYY')
order by s.stationcode , cname
```
![Image](https://github.com/user-attachments/assets/578f57c8-a1b0-499e-b5ba-2ce8fd25c3f0)

**Comment by clezag** [comment]
@rcavaliere what's the status on this issue? Is there something still open from our side?

**Comment by rcavaliere** [comment]
@clezag we need to check if all traffic stations have the right sensor_type information provided, so that we can check the correct calculation of the emissions

**Comment by clezag** [comment]
@rcavaliere 

My plan was to

- delete all data that has been written by the pollution_v2 elaboration since **2023** (RIGHT? from my understanding we never did that in production) 
- relaunch the elaboration with starting date 01.01.**2023** with the updated values for 2023 and 2024.

For the CCTV cameras, do we really have to do handle them differently?
It's my understanding that the pollution elaboration only calculates data on validated stations.  
If the data from before 10.07.2024 is unreliable, it should not pass validation, and therefore not be elaborated.
And if it passes, it's reliable so why not calculate?

Is there a specific reason why you don't want to recalculate from 2023?
If there is, we could add the logic just into the validator and always set validate to false if `type = camera and date < 10.07.2024` and consider them all invalid

**Comment by clezag** [comment]
@rcavaliere I can't tell yet, just letting it run through as is will probably take a week+, but I'm trying to work out some optimizations, and monitor the data while it's being written to catch problems early and restart if necessary.

**Comment by rcavaliere** [comment]
@clezag thanks for the update, looks promising!
In the test data you posted I see a lot of invalidations, that could be indeed possibìle (to be verified with CISMA).
Before merging #49, I would check if the issue #50 is solved (I don't think so, since in the PR description the solution of such an issue is not present). Can you please write to Marco A. and the colleagues of u-hopper asking to solve this? After that I will proceed with this operation and let CISMA control with me if the emissions calculation in our testing environment is OK or not.

**Comment by rcavaliere** [comment]
@clezag just written to Afliant, let's wait for a feedback

**Comment by rcavaliere** [comment]
@clezag yes, I would go in that direction. The recalculation should be then configured so to have the final desired result, as described in the user story and also in this comment: https://github.com/noi-techpark/opendatahub-elaborations/issues/38#issuecomment-2505347330 In case we need external support from Afliant and CISMA, please let me know!

**Comment by rcavaliere** [comment]
@clezag I would suggest to proceed now with the operation with deletion and re-elaboration. Maybe it is worth to make the elaboration launch / configuration with Afliant, so that you also better understand how to manage airflow. Once we have this consolidated in testing, I would like to let CISMA check the results, before putting everything in production

**Comment by clezag** [comment]
@rcavaliere currently the new DAGs fail because they use newly introduced services which we don't have the build specs of. See #52

**Comment by rcavaliere** [comment]
@clezag today CISMA has explained me an additional significant improvement they would like to introduce in the elaboration logic so to make much better estimations of the emissions. From an informatics point of view, we should include, similarly as for the emission factors (see file "fc_xxxx"), one file for each copert55.db, as the one in annex. My proposal, since we are putting the hand for the last time here, is to restart the computation from 1.1.2024 by considering this file instead of the previous copert55.db file. 

With CISMA we should then find a way to update the calculation logic, since every year they will provide new input files to be considered in the calculation. The simplest way to do this is to delete the elaborations computed up to that moment and remake the calculations running since 1.1.202x.

One last thing: for the testing environment is not necessary to remake the calculation, just update the file and the reference and stop.

Let me know if you can don this on your own or we need the support of U-Hopper who implemented the code.

[copert55_2024.db.zip](https://github.com/user-attachments/files/22043533/copert55_2024.db.zip)

**Comment by rcavaliere** [comment]
@clezag but did you change the reference in the source code? It seems to me that it is looking for the old file from the log you shared.

**Comment by clezag** [comment]
@rcavaliere There is an issue with duplicate records that causes the elaboration to not advance. Still not sure about the exact extent, but it's somehow related to the stationcode migration, I'm working on cleaning it up

**Comment by clezag** [comment]
@rcavaliere I've cleaned up the duplicate records (always retaining the most recently written one - we couldn't go on provenance because they even had duplicates within the same provenance).

Now I see the validation job succeeding for the first time, will check back in a bit how it's doing. From past experience it takes quite some time to catch up

**Comment by rcavaliere** [comment]
@clezag yes I was checking in these days. Yes, we are nearly arrived with the historical calculation, let's what happens when we reach the status "now".
